pooch/downloaders.py

```diff
@@ -1034,20 +1034,31 @@ def download_url(self, file_name):
         download_url : str
             The HTTP URL that can be used to download the file.
         """
-
         parsed = parse_url(self.archive_url)
-
-        # Iterate over the given files until we find one of the requested name
-        for filedata in self.api_response.json()["data"]["latestVersion"]["files"]:
-            if file_name == filedata["dataFile"]["filename"]:
-                return (
-                    f"{parsed['protocol']}://{parsed['netloc']}/api/access/datafile/"
-                    f":persistentId?persistentId={filedata['dataFile']['persistentId']}"
-                )
-
-        raise ValueError(
-            f"File '{file_name}' not found in data archive {self.archive_url} (doi:{self.doi})."
-        )
+        response = self.api_response.json()
+        files = {
+            file["dataFile"]["filename"]: file["dataFile"]
+            for file in response["data"]["latestVersion"]["files"]
+        }
+        if file_name not in files:
+            raise ValueError(
+                f"File '{file_name}' not found in data archive "
+                f"{self.archive_url} (doi:{self.doi})."
+            )
+        # Generate download_url using persistentId or file id
+        persistent_id = files[file_name]["persistentId"]
+        if persistent_id:
```

Comment:
The file ID will always be there. It's the primary key in the database.

The persistent ID (DOI or Handle) is optional so you can't rely on it being there. I would simply avoid even checking for it if all you want to do is download the file.

Reply:
I see. I thought the `persistentId` is always there, but it's empty if the files doesn't have one. Sorry for the misunderstanding.

If that's the case, you're right, we shouldn't assume that `persistentId` will always be there. I'll change the `if` statement then.

BTW, do you have any example where the `persistentId` is not even included in the response (I'm thinking for testing purposes)?

Reply:
I would suggest changing your strategy to this:

- Only check for `id`

Simple! 😄